### PR TITLE
Serialize reminders as protobuf in state store

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
Part of #6121 

Changes how reminders are serialized in the state store, switching from the current JSON serialization to protobuf. Using protobuf is more efficient, not just because the serialization itself is faster, but because the payload is more "compressed" and so the I/O cost (the biggest issue today) is reduced.

Results from the actor_reminder perf test (using Postgres) (higher is better):

- [JSON](https://github.com/dapr/dapr/actions/runs/6708202961): 37.80 QPS
- [Protobuf](https://github.com/dapr/dapr/actions/runs/6710588956): 58.96 QPS (+56%)

Workflow perf test (TestSeriesWorkflowWithMaxVUs - Req Duration P90) (lower is better):

- [JSON](https://github.com/dapr/dapr/actions/runs/6708202961): 123.4s
- [Protobuf](https://github.com/dapr/dapr/actions/runs/6710609050): 54.5s (-55%)

This is the first change that leverages #6838 for real to preserve backwards-compatibility and to be able to function in a cluster with mixed Dapr versions

- Actor API level has been bumped to 20 (was 10 in 1.12)
- If the cluster API level is 20 or higher, reminders are serialized as protobuf. This means that if all actor hosts in the cluster are running on "vNext" (assuming 1.13), then reminders are serialized to protobuf.
- If the actors API level in the cluster is < 20 (either because there's at least 1 host running 1.12 or lower, or because the admin explicitly set the maximum API version to < 20 in the Helm chart), then reminders are serialized as JSON
- It's always possible to parse reminders that are serialized as JSON, regardless of the current API level

TODO:

- [ ] Figure out how to write a test for ensuring that this can read/write JSON depending on the API level in the cluster

### CODE REVIEW

File: dapr/proto/internals/v1/reminders.proto [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+// Reminder represents a reminder that is stored in the Dapr actor state store.
+message Reminder {
+  string actor_id = 1;
+  string actor_type = 2;
+  string name = 3;
+  bytes data = 4;
+  string period = 5;
+  google.protobuf.Timestamp registered_time = 6;
+  string due_time = 7;
```
Comment (cgillum) [inline dapr/proto/internals/v1/reminders.proto:30]: Why `string` and not `google.protobuf.Timestamp` for `due_time`?

Comment (ItalyPaleAle) [reply dapr/proto/internals/v1/reminders.proto:30]: it's the value as sent by the user, and it could be a duration such as "2m". A bit of a legacy thing.

**Comment by yaron2** [comment]
Can attest personally changing JSON to proto several years back has improved Redis throughput x10 on a project I worked on

**Comment by artursouza** [comment]
I really like this change. It is a quick life improvement to users before we can offer a v2 experience for reminders.

**Review by mukundansundar — COMMENTED** [review]
the change is something that will be extremely useful. Please add version skew tests for this feature.

**Comment by ItalyPaleAle** [comment]
> the change is something that will be extremely useful. Please add version skew tests for this feature.

I'll see what I can do after I'm back next week.

The current setup for version skew tests doesn't allow testing for something like this. We may need some creative thinking.

**Comment by ItalyPaleAle** [comment]
I have added integration tests that verify the behavior. In particular:

1. That daprd can read reminders serialized as protobuf or JSON (this works as "upgrade test")
2. That when the Actors API level in the cluster is >= 20, reminders are stored as protobuf
3. That when the Actors API level in the cluster is < 20, reminders are stored as JSON

@mukundansundar PTAL

**Comment by JoshVanL** [comment]
@ItalyPaleAle @mukundansundar  
friendly reminder that actor/placement PRs are blocked until https://github.com/dapr/dapr/issues/7250 is resolved

**Comment by ItalyPaleAle** [comment]
This PR actually includes a fix for the error in 7250. Not sure why it does appear only on MacOS. But i fixed it here

File: tests/integration/framework/process/daprd/options.go [hunk 2]
```diff
@@ -136,6 +146,37 @@ func WithResourceFiles(files ...string) Option {
...
+    - name: connectionString
+      value: 'file:` + dbPath + `'
+    - name: actorStateStore
+      value: 'true'
+`)
+
+	// Connect to the database
+	conn, err := sql.Open("sqlite", "file://"+dbPath+"?_txlock=immediate&_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)")
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to open connection to SQLite database: %w", err)
```
Comment (JoshVanL) [inline tests/integration/framework/process/daprd/options.go:175]: Don't return error, and use `t` instead.

File: tests/integration/suite/actors/reminders/serialization.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	"github.com/dapr/dapr/tests/integration/framework/process/daprd"
+	prochttp "github.com/dapr/dapr/tests/integration/framework/process/http"
+	"github.com/dapr/dapr/tests/integration/framework/process/placement"
+	"github.com/dapr/dapr/tests/integration/framework/util"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(serializationProtobuf))
+	suite.Register(new(serializationJSON))
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization.go:43]: Put these into separate files `reminders/serialisation/json.go` `reminders/serialisation/protobuf`

File: tests/integration/framework/process/daprd/options.go [hunk 4]
```diff
@@ -136,6 +146,37 @@ func WithResourceFiles(files ...string) Option {
...
+  version: v1
+  metadata:
+    - name: connectionString
+      value: 'file:` + dbPath + `'
+    - name: actorStateStore
+      value: 'true'
+`)
+
+	// Connect to the database
+	conn, err := sql.Open("sqlite", "file://"+dbPath+"?_txlock=immediate&_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)")
```
Comment (JoshVanL) [inline tests/integration/framework/process/daprd/options.go:173]: Connection should be closed when test is cleaned up.

Also consider moving SQLite to a separate framework process. We should only ever be retuning `*Options` from option funcs, so this suggests this setup isn't quite right.

File: tests/perf/actor_reminder/actor_reminder_test.go [hunk 5]
```diff
@@ -39,7 +39,7 @@ const (
 	appName         = "perf-actor-reminder-service"
 
 	// Target for the QPS
-	targetQPS = 50
+	targetQPS = 56
```
Comment (JoshVanL) [inline tests/perf/actor_reminder/actor_reminder_test.go:42]: Why?

File: tests/integration/suite/actors/reminders/serialization.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	place   *placement.Placement
+	db      *sql.DB
+}
+
+func (i *serializationProtobuf) Setup(t *testing.T) []framework.Option {
+	// Init placement with no maximum API level
+	i.place = placement.New(t)
+
+	// Create a SQLite database
+	sqliteOpt, db, err := daprd.WithSQLiteStateStore(t)
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization.go:62]: `WithSQLightActorStateStore` should be receiving an SQLite process.

File: tests/integration/suite/actors/reminders/serialization.go (hunk 3 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization.go:43]: A lot of the code is shared

**Comment by ItalyPaleAle** [comment]
> @ItalyPaleAle it's not clear to me why this should be fixing the actor tests, can you explain further?

I can't link right now, on my phone. It's L73 in rebalancing.go that fixes it by setting a maximum version for actors api level.

I stumbled upon that by accident and fixed it in this PR before I saw your issue with the log indicating that failure.

It may appear just on Mac due to timings.

File: tests/integration/framework/process/daprd/options.go (hunk 4 above)
Comment (ItalyPaleAle) [reply tests/integration/framework/process/daprd/options.go:173]: No need to close, the connection is automatically closed when the object is deallocated. It's also a temporary file so doesn't matter anyways.

As for a separate process, SQLite is not a process that is spawned, so not sure I understand what you mean.

File: tests/perf/actor_reminder/actor_reminder_test.go (hunk 5 above)
Comment (ItalyPaleAle) [reply tests/perf/actor_reminder/actor_reminder_test.go:42]: Because we can now :) the perf has improved a lot so setting a new baseline.

File: tests/integration/suite/actors/reminders/serialization.go (hunk 6 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization.go:62]: There's no process being spawned.

File: tests/integration/framework/process/daprd/options.go (hunk 4 above)
Comment (JoshVanL) [reply tests/integration/framework/process/daprd/options.go:173]: We want to make this a separate process so that it itself can be passed to other options. The SQLite process would then expose a `DBFile() string` method which can be used by receiving Option funcs.

I'd rather we did close the connection for completeness. We will then know for sure all resources are cleaned up on test completion.

Comment (JoshVanL) [reply tests/integration/framework/process/daprd/options.go:173]: To be clear here- an Integration framework process doesn't refer to a Linux process (although it can), it instead refers to a long running test dependency which has a Setup, Run, and exposes runtime config which can be used by other processes or the test itself. In this case, the SQLite will expose the db file created, and/or a connection to that file.

**Comment by ItalyPaleAle** [comment]
@JoshVanL i've updated the tests per your feedback

File: tests/integration/suite/actors/reminders/rebalancing.go [hunk 7]
```diff
@@ -241,15 +229,15 @@ func (i *rebalancing) Run(t *testing.T, ctx context.Context) {
 	}
 }
 
-type httpServer struct {
+type rebalancingHTTPServer struct {
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/rebalancing.go:232]: We don't need this change because `rebalancing` is inferred through the test (package) name.

File: tests/integration/suite/actors/reminders/serialization_json.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+	// Check the data in the SQLite database
+	// The value must begin with `[{`, which indicates it was serialized as JSON
+	storedVal := serializationLoadRemindersFromDB(t, ctx, i.db.GetConnection())
+	assert.Truef(t, strings.HasPrefix(storedVal, "[{"), "Prefix not found in value: '%v'", storedVal)
+
+	// Ensure the reminder was invoked at least once
+	assert.Eventually(t, func() bool {
+		return i.handler.remindersInvokeCount.Load() > 0
+	}, time.Second, 10*time.Millisecond, "Reminder was not invoked at least once")
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization_json.go:104]: A second seems very low, are we worried about this?

**Review by JoshVanL — CHANGES_REQUESTED** [review]
@ItalyPaleAle the actor integration tests still don't seem happy.

File: tests/integration/suite/actors/reminders/rebalancing.go (hunk 7 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/rebalancing.go:232]: There is another HTTP server in the package.

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	for _, fopt := range fopts {
+		fopt(&s.opts)
+	}
+
+	// Create a SQLite database in the test's temporary directory
+	tmpDir := t.TempDir()
+	s.dbPath = filepath.Join(tmpDir, "test-data.db")
+	// t.Logf("Storing SQLite database at %s", s.dbPath)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:63]: It's very helpful for debugging the test.

File: tests/integration/framework/process/daprd/options.go [hunk 10]
```diff
@@ -13,7 +13,13 @@ limitations under the License.
 
 package daprd
 
-import "github.com/dapr/dapr/tests/integration/framework/process/exec"
+import (
+
+	// Blank import for the sqlite driver
+	_ "modernc.org/sqlite"
+
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/daprd/options.go:20]: Without this it won't work

File: tests/integration/suite/actors/reminders/serialization_json.go (hunk 8 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization_json.go:104]:
```suggestion
	}, 5*time.Second, 10*time.Millisecond, "Reminder was not invoked at least once")
```

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+	// Connect to the database
+	// We do this here because the test's init code may need to access the database before starting the test
+	if s.opts.connect {
+		var err error
+		s.conn, err = sql.Open("sqlite", "file://"+s.dbPath+"?_txlock=immediate&_pragma=busy_timeout(5000)&_pragma=journal_mode(WAL)")
+		require.NoError(t, err, "Failed to connect to SQLite database")
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:70]: That's why there's `s.opts.connect`. It either needs the connection for the entire test, or not. It should be fine.

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+INSERT INTO metadata VALUES('migrations','1');
+CREATE TABLE state (
+  key TEXT NOT NULL PRIMARY KEY,
+  value TEXT NOT NULL,
+  is_binary BOOLEAN NOT NULL,
+  etag TEXT NOT NULL,
+  expiration_time TIMESTAMP DEFAULT NULL,
+  update_time TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
+);`)
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:110]: It should only be used in some edge cases, such as when the database needs to be pre-populated with data.

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	for k, v := range s.opts.metadata {
+		c.Spec.Metadata = append(c.Spec.Metadata, commonapi.NameValuePair{
+			Name:  k,
+			Value: toDynamicValue(v),
+		})
+	}
+
+	enc, _ := json.Marshal(c)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:140]: This method doesn't accept a `t`. And this can't really fail because we only pass values that can be serialized as JSON.

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	return string(enc)
+}
+
+func toDynamicValue(val string) commonapi.DynamicValue {
+	j, _ := json.Marshal(val)
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:145]: The caller of this method doesn't have `t`. And strings never fail to be serialized to JSON

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go:63]: I prefer not to include these comment out debug lines as it increases noise in the code.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go:70]: The `conn` here is a runtime test dependency and does not belong in the `New` function. Please look at other framework processes to see how `New` is used to instantiate options and configs. Runtime dependencies are created through `Run`. Ports are the only special case I can think of because of the nature of needing to both retrieve a reserved port from `0` and needing to squat that port till test execution. We don't have that requirement here.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go:140]: The method can be updated to accept a `t *testing.T` and is proper to always check errors. The contents may be updated in future, error, and cause headaches for the programmer.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go:145]: Same as above, please check all errors.

File: tests/integration/suite/actors/reminders/serialization_json.go [hunk 15]
```diff
@@ -0,0 +1,105 @@
+/*
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/serialization_json.go:1]: This is not a resolved comment.

File: tests/integration/suite/actors/reminders/serialization_protobuf.go [hunk 16]
```diff
@@ -0,0 +1,116 @@
+/*
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/serialization_protobuf.go:1]: Same here.

File: tests/integration/suite/actors/reminders/rebalancing.go (hunk 7 above)
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/rebalancing.go:232]: This will not be the case with `serialization` being made a subpackage.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:110]: No, we can't do that in `Run` because it's executed in parallel with daprd starting. This must be done before daprd starts up

**Comment by JoshVanL** [comment]
@ItalyPaleAle it shouldn't be the case that we need _more_ resources for the runners when things were working before. This suggests to me that resources are not being cleaned up between test cases, or a test case is completely de-funked.

File: tests/integration/suite/actors/reminders/serialization_json.go (hunk 8 above)
Comment (ItalyPaleAle) [inline tests/integration/suite/actors/reminders/serialization_json.go]:
```suggestion
	}, 5*time.Second, 10*time.Millisecond, "Reminder was not invoked at least once")
```

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:140]: Our metadata only accepts strings so really this can never fail. Let's keep it simple :)

File: tests/integration/suite/actors/reminders/serialization_protobuf.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization_protobuf.go:1]: Why does it need a sub-package? It's still reminders stuff

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:63]: I heavily relied on this log during development, to access the DB and see the data inside...

File: tests/integration/suite/actors/reminders/serialization_protobuf.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/serialization_protobuf.go:1]: Yes, `serialization` is a sub topic of reminders- like `reminders` being a sub topic of `actors`. Package namespacing ensures better code organization and removes code smells like `serializationHTTPServer` & `rebalancingHTTPServer`.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:70]: Ok

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go:110]: Sorry I misspoke. Now that I've more time I went back and looked again at what the issue was.

We must create the tables before calling `Run`, because we must write data in the database before calling `Run`. Otherwise, we'd have to populate the database in the **test's** `Run` method, but at that point daprd is already running.

File: tests/integration/suite/actors/reminders/serialization/json.go [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/dapr/dapr/tests/integration/framework/process/daprd"
+	prochttp "github.com/dapr/dapr/tests/integration/framework/process/http"
+	"github.com/dapr/dapr/tests/integration/framework/process/placement"
+	"github.com/dapr/dapr/tests/integration/framework/process/sqlite"
+	"github.com/dapr/dapr/tests/integration/framework/util"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(serializationJSON))
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/json.go:37]: `serialization` is part of the package (and test name) so inferred. 

```suggestion
	suite.Register(new(json)
```

File: tests/integration/suite/actors/reminders/serialization/json.go [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// serializationJSON tests:
+// - That reminders are serialized to JSON when the Actors API level in the cluster is < 20
+type serializationJSON struct {
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/json.go:42]:
```suggestion
type json struct {
```

File: tests/integration/suite/actors/reminders/serialization/json.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/dapr/dapr/tests/integration/framework/process/sqlite"
+	"github.com/dapr/dapr/tests/integration/framework/util"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(serializationJSON))
+}
+
+// serializationJSON tests:
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/json.go:40]:
```suggestion
// json tests:
```

File: tests/integration/suite/actors/reminders/serialization/json.go [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+// - That reminders are serialized to JSON when the Actors API level in the cluster is < 20
+type serializationJSON struct {
+	daprd   *daprd.Daprd
+	srv     *prochttp.HTTP
+	handler *httpServer
+	place   *placement.Placement
+	db      *sqlite.SQLite
+}
+
+func (i *serializationJSON) Setup(t *testing.T) []framework.Option {
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/json.go:50]:
```suggestion
func (j *json) Setup(t *testing.T) []framework.Option {
```

File: tests/integration/suite/actors/reminders/serialization/protobuf.go [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/dapr/dapr/tests/integration/framework/process/daprd"
+	prochttp "github.com/dapr/dapr/tests/integration/framework/process/http"
+	"github.com/dapr/dapr/tests/integration/framework/process/placement"
+	"github.com/dapr/dapr/tests/integration/framework/process/sqlite"
+	"github.com/dapr/dapr/tests/integration/framework/util"
+	"github.com/dapr/dapr/tests/integration/suite"
+)
+
+func init() {
+	suite.Register(new(serializationProtobuf))
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/protobuf.go:38]: Also here.

```suggestion
	suite.Register(new(protobuf))
```

File: tests/integration/suite/actors/reminders/serialization/protobuf.go [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+	// This will assert the ability to "upgrade" from JSON to Protobuf
+	queryCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	now := time.Now().UTC().Format(time.RFC3339)
+	_, err := i.db.GetConnection(t).ExecContext(queryCtx, fmt.Sprintf(`
+INSERT INTO state VALUES
+  ('actors||myactortype','[{"registeredTime":"%[1]s","period":"2m","actorID":"myactorid","actorType":"myactortype","name":"oldreminder","dueTime":"0"}]',0,'e467f810-4e93-45ed-85d9-e68d9fc7af4a',NULL,'%[1]s'),
+  ('actors||myactortype||metadata','{"id":"00000000-0000-0000-0000-000000000000","actorRemindersMetadata":{"partitionCount":0}}',0,'e82c5496-ae32-40a6-9578-6a7bd84ff331',NULL,'%[1]s');
+`, now))
+	require.NoError(t, err)
```
Comment (JoshVanL) [inline tests/integration/suite/actors/reminders/serialization/protobuf.go:81]: This doesn't belong here- the code smell is the `context.Background()`.

This should be moved to a `WithRunExec` option which is executed on the process `Run`. This will then also have the proper test context a available.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go]: Daprd is not running at this point.

File: tests/integration/framework/process/sqlite/sqlite.go [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+	// Create a SQLite database in the test's temporary directory
+	tmpDir := t.TempDir()
+	s.dbPath = filepath.Join(tmpDir, "test-data.db")
+	t.Logf("Storing SQLite database at %s", s.dbPath)
+
+	return s
+}
+
+func (s *SQLite) Run(t *testing.T, ctx context.Context) {
+	// Nothing here
```
Comment (JoshVanL) [inline tests/integration/framework/process/sqlite/sqlite.go:70]: This is where the SQLite Exec setup belongs.

**Review by JoshVanL — CHANGES_REQUESTED** [review]

**Comment by JoshVanL** [comment]
> > @ItalyPaleAle it shouldn't be the case that we need _more_ resources for the runners when things were working before. This suggests to me that resources are not being cleaned up between test cases, or a test case is completely de-funked.
> 
> I don't know what to say... Maybe more tests now are using more CPU and we're exhausting resources?

The error is because of the actor rebalancing test. You can see this here https://github.com/JoshVanL/dapr/actions/runs/7060286640/job/19219678365?pr=28

File: tests/integration/suite/actors/reminders/serialization/json.go (hunk 17 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization/json.go:37]: That would conflict with the json package

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go]: When the test is Run-ing, daprd is not??

**Comment by yaron2** [comment]
@ItalyPaleAle once you address existing comments I think we'll be good to merge, brief review LGTM.

File: tests/integration/suite/actors/reminders/serialization/json.go (hunk 18 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization/json.go:42]: that would conflict with the `json` package

File: tests/integration/suite/actors/reminders/serialization/protobuf.go (hunk 21 above)
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/serialization/protobuf.go:38]: Because the other test is called `serializationJSON` (and can't be `json` as it would conflict with the serialization package), this should stay as `serializationProtobuf`

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/framework/process/sqlite/sqlite.go]: I don't think there's anything I can do here as we need to run the queries before daprd starts.

Regardless of whether processes are started in sequence or parallel, the test's `Run` method is invoked after all processes have started, so we cannot execute queries to write data in that method.

Please confirm and I'll resolve the conversation.

**Comment by ItalyPaleAle** [comment]
The second bug in the reminders rebalancing IT should be fixed by d25832b

File: tests/integration/suite/actors/reminders/serialization/json.go (hunk 17 above)
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/serialization/json.go:37]: We can use an identifier for the `encoding/json` package such as `encodingjson encoding/json`. I think this is better than having stuttering in the test name.

File: tests/integration/suite/actors/reminders/serialization/protobuf.go (hunk 21 above)
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/serialization/protobuf.go:38]: Please see above.

File: tests/integration/framework/process/sqlite/sqlite.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/framework/process/sqlite/sqlite.go]: Sorry, I don't know what you mean and think there is some confusion on the intention of a `suite.Case`, the framework `Process` and the order things are executed.

The test `Case` is the main test definition, whose struct name and package tree derive the Go test name. Its `Setup` function is always run first, and is where test processes (test dependencies) are defined. Next, the framework will execute the `Run` function of the test case process dependencies in order. This is where the process executes the actual dependency- this could be a binary process, in memory server, writing to a database w.e.
The test `Case` Run is then executed where the test invocation and checking code lives. Finally the processes are then `Cleanup`ed in reverse order to the `Run` order- it is important that processes ensure all resources are released here.

Below is the patch file for these changes showing how the integration test process is designed to be structured.

[sqlite.patch.txt](https://github.com/dapr/dapr/files/13599591/sqlite.patch.txt)

**Comment by ItalyPaleAle** [comment]
Looks like the race condition is another bug in that test case (only in test code) that predates this PR. Fix doesn't seem immediate, but I can look into fixing that in a separate PR.

The remaining comments are about nits on test code, including naming, and can be fixed separately if needed. Looking forward getting this PR merged to fix the perf and integration tests failing in the master branch for many weeks :)

File: tests/integration/suite/actors/reminders/rebalancing.go [hunk 24]
```diff
@@ -419,14 +406,19 @@ func (i *rebalancing) reportStatusToPlacement(ctx context.Context, stream placem
...
+				}
+				return
 			}
-			if o.GetOperation() == "update" {
+			if o.GetOperation() == "update" && !sent {
 				errCh <- nil
-				return
+				sent = true
 			}
 		}
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply tests/integration/suite/actors/reminders/rebalancing.go:423]: I do not get a race condition for this test on master.

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply tests/integration/suite/actors/reminders/rebalancing.go:423]: The race condition is due to concurrent `recv` and `send` in the gRPC channel. Not sure why it doesn't appear on master (maybe because the background goroutine was terminated on the first message?) but that's not new code. Fixing that requires creating 2 separate streams and it's a moderately-sized code change. Can be fixed later

**Comment by artursouza** [comment]
> Looks like the race condition is another bug in that test case (only in test code) that predates this PR. Fix doesn't seem immediate, but I can look into fixing that in a separate PR.
> 
> The remaining comments are about nits on test code, including naming, and can be fixed separately if needed. Looking forward getting this PR merged to fix the perf and integration tests failing in the master branch for many weeks :)

There might be 2 different issues here, so I don't want to dismiss the race condition in this PR or a bug in the test code.

On the other hand, we should raise our bar in pull requests because our users (and us) are paying a price for that. I suggest we move slower but get the PRs to a mergeable state and avoid "fix later" as much as possible. It is better to move slower in the right direction than fast in the wrong one.

**Review by yaron2 — APPROVED** [review]

**Comment by yaron2** [comment]
Merging this as the race condition is scoped to test code and we're getting close to code freeze. We have a plan with @JoshVanL to follow up with a fix.
